### PR TITLE
feat!: Remove valid-jsdoc and require-jsdoc

### LINKED ISSUE #15820: Remove `require-jsdoc` and `valid-jsdoc` (and `doctrine`)
### ESLint version

v8.14.0

### What problem do you want to solve?

The `require-jsdoc` and `valid-jsdoc` rules were deprecated [a good while ago](https://eslint.org/blog/2018/11/jsdoc-end-of-life), but despite having [replacements in `eslint-plugin-jsdoc`](https://github.com/gajus/eslint-plugin-jsdoc/wiki/Comparison-with-deprecated-JSdoc-related-ESLint-rules), a major version change hasn't removed them.

### What do you think is the correct solution?

For the next major version, I think we should remove the rules (and the `doctrine` dependency which is only being used by these rules).

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

Remove rule

#### What changes did you make? (Give an overview)

1. Removed `valid-jsdoc` rule, tests, and docs
2. Updated rule deprecation policy

TODO: Determine if we want to remove `require-jsdoc` too.

Fixes #15820

### CODE REVIEW

**Comment by mdjermanovic** [comment]
@nzakas you probably intended to push https://github.com/eslint/eslint/pull/17694/commits/3deb185fe7435fef93d33a70708999861cf77ef2 to the branch of https://github.com/eslint/eslint/pull/17696 instead of here.

File: conf/rule-type-list.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
         { "removed": "space-in-brackets", "replacedBy": ["object-curly-spacing", "array-bracket-spacing"] },
         { "removed": "space-return-throw-case", "replacedBy": ["keyword-spacing"] },
         { "removed": "space-unary-word-ops", "replacedBy": ["space-unary-ops"] },
-        { "removed": "spaced-line-comment", "replacedBy": ["spaced-comment"] }
+        { "removed": "spaced-line-comment", "replacedBy": ["spaced-comment"] },
+        { "removed": "valid-jsdoc", "replacedBy": [] }
```
Comment (mdjermanovic) [inline conf/rule-type-list.json:27]: We should also add `require-jsdoc` to this list.

File: docs/src/rules/require-jsdoc.md
Comment (mdjermanovic) [inline docs/src/rules/require-jsdoc.md:1]: We're not removing docs for removed rules. We're just updating it with the info that the rule has been removed.

For example:

https://eslint.org/docs/latest/rules/generator-star

Comment (nzakas) [reply docs/src/rules/require-jsdoc.md:1]: Ah! Pretty sure we had a stronger visual treatment than that, though. I'll check on that too.

Comment (mdjermanovic) [reply docs/src/rules/require-jsdoc.md:1]: Yeah, it seems that on the old site that `(removed)` was converted to an X.

https://archive.eslint.org/docs/rules/generator-star

File: docs/src/use/rule-deprecation.md [hunk 2]
```diff
@@ -7,10 +7,12 @@ Balancing the trade-offs of improving a tool and the frustration these changes c
 
 The ESLint team is committed to making upgrading as easy and painless as possible. To that end, the team has agreed upon the following set of guidelines for deprecating rules in the future. The goal of these guidelines is to allow for improvements and changes to be made without breaking existing configurations.
 
-* Rules will never be removed from ESLint.
+* Rules will never be removed from ESLint unless one of the following is true:
+    * The rule has been replaced by another core rule
```
Comment (mdjermanovic) [inline docs/src/use/rule-deprecation.md:11]:
```suggestion
    * The rule has been replaced by another core rule.
```

File: tests/lib/cli-engine/cli-engine.js [hunk 3]
```diff
@@ -1774,7 +1772,7 @@ describe("CLIEngine", () => {
             engine = new CLIEngine({
                 cwd: originalDir,
                 useEslintrc: false,
-                rules: { indent: 1, "valid-jsdoc": 0, "require-jsdoc": 0 }
+                rules: { indent: 1, "callback-return": 0 }
```
Comment (mdjermanovic) [inline tests/lib/cli-engine/cli-engine.js:1775]: This line is causing merge conflicts. The fixed line should be `rules: { eqeqeq: 1, "callback-return": 0 }` (`indent` was replaced here with `eqeqeq` in https://github.com/eslint/eslint/pull/17696).

File: tests/lib/eslint/eslint.js [hunk 4]
```diff
@@ -1762,7 +1760,7 @@ describe("ESLint", () => {
                 cwd: originalDir,
                 useEslintrc: false,
                 overrideConfig: {
-                    rules: { indent: 1, "valid-jsdoc": 0, "require-jsdoc": 0 }
+                    rules: { indent: 1, "callback-return": 0 }
```
Comment (mdjermanovic) [inline tests/lib/eslint/eslint.js:1763]: Same as https://github.com/eslint/eslint/pull/17694#discussion_r1398929976

File: tests/lib/eslint/flat-eslint.js [hunk 5]
```diff
@@ -1956,7 +1954,7 @@ describe("FlatESLint", () => {
                     cwd: originalDir,
                     overrideConfigFile: true,
                     overrideConfig: {
-                        rules: { indent: 1, "valid-jsdoc": 0, "require-jsdoc": 0 }
+                        rules: { indent: 1, "callback-return": 0 }
```
Comment (mdjermanovic) [inline tests/lib/eslint/flat-eslint.js:1957]: Same as https://github.com/eslint/eslint/pull/17694#discussion_r1398929976

**Comment by mdjermanovic** [comment]
There are also merge conflicts in `docs/src/_data/rules.json`. I think it would be better to undo changes in autogenerated files to reduce possible merge conflicts.

File: tests/lib/eslint/flat-eslint.js [hunk 6]
```diff
@@ -1966,7 +1964,7 @@ describe("FlatESLint", () => {
                     cwd: originalDir,
                     overrideConfigFile: true,
                     overrideConfig: {
-                        rules: { eqeqeq: 1, "valid-jsdoc": 0, "require-jsdoc": 0 }
+                        rules: { eqeqeq: 1, "require-jsdoc": 0 }
```
Comment (mdjermanovic) [inline tests/lib/eslint/flat-eslint.js:1967]:
```suggestion
                        rules: { eqeqeq: 1, "callback-return": 0 }
```

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Comment by mdjermanovic** [comment]
There are merge conflicts because the two removed test files have changed in the meantime.

File: tests/lib/eslint/flat-eslint.js [hunk 7]
```diff
@@ -1965,7 +1963,7 @@ describe("FlatESLint", () => {
...
```
Comment (mdjermanovic) [inline tests/lib/eslint/flat-eslint.js:1966]:
```suggestion
                        rules: { eqeqeq: 1, "callback-return": 0 }
```

**Review by mdjermanovic — COMMENTED** [review]
Looks like https://github.com/eslint/eslint/commit/7a7944505dff66f0452395caef8dc8e6c5ce96a4 was lost in rebase.

**Comment by nzakas** [comment]
Thanks for that. These conflicts were getting messy.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**